### PR TITLE
support custom prompts in autocomplete

### PR DESCRIPTION
**Pull request by yen-tt**
this pr provide support for user to add custom autocomplete prompts in addition to the autocomplete results from core. 

User can either specify an array of string for custom prompts through searchbar.autocomplete config (`ANSWERS.addComponent('SearchBar', {..., autocomplete: {customPrompts: ['prompt']} }`) or use `setCustomPrompts` (use in theme). In autocomplete's setState, custom prompts will transform into right data format and merge with core results. This is then set in storage to trigger a new state update with the combined result. Custom prompts appear on empty search and persist when user type similar substring as the prompt.

J=SLAP-1803
TEST=manual

- serve page and add SearchBar with customPrompts configuration, with and without hardcoded prompts from backend. See that custom prompts appear as expected. 
  - Type same starting substrings and see that custom prompts still display (levenshtein match) 
  - Type and remove input, see that custom prompts appear again.
- Hook SDK changes to theme. spin up test site and see that custom prompts appear as expected

### CODE REVIEW

File: src/ui/components/search/autocompletecomponent.js [hunk 1]
```diff
@@ -179,6 +186,15 @@ export default class AutoCompleteComponent extends Component {
       this._isOpen = true;
     }
 
+    if (this.customPrompts && !queryInputEl.value && data.sections) {
```
Comment (tmeyer2115) [inline src/ui/components/search/autocompletecomponent.js]: Does this conditional mean that we don't show the custom prompts if there's some text in the `input`? What if the text in the input matches what's in some of the prompts?

File: src/ui/components/search/autocompletecomponent.js [hunk 2]
```diff
@@ -358,6 +374,23 @@ export default class AutoCompleteComponent extends Component {
     this._isVoiceSearchActive = value;
   }
 
+  setCustomPrompts (prompts) {
+    this.customPrompts = prompts;
+  }
+
+  combineResultsWithCustomPrompts (data) {
+    const newSections = data.sections.filter(result => result.resultsCount !== 0);
```
Comment (tmeyer2115) [inline src/ui/components/search/autocompletecomponent.js]: Small nit: Could we break this method up? Maybe have a `createSectionFromCustomPrompts` and then the combine method?

File: src/ui/components/search/autocompletecomponent.js (hunk 1 above)
Comment (yen-tt) [reply src/ui/components/search/autocompletecomponent.js]: Yes. Product was ok with custom prompts disappearing when user start typing. Although, they also ask to match the behavior of the hardcoded prompts that user can configure in the backend, which does include the prompts if input have some sort of text matching. I can add that check using the levenshtein functions we used in headless

File: src/ui/components/search/autocompletecomponent.js (hunk 2 above)
Comment (yen-tt) [reply src/ui/components/search/autocompletecomponent.js]: sure thing

File: src/ui/components/search/autocompletecomponent.js [hunk 3]
```diff
@@ -179,6 +187,18 @@ export default class AutoCompleteComponent extends Component {
       this._isOpen = true;
     }
 
+    // should show IF it's empty search OR levenstien match on a non empty search
```
Comment (oshi97) [inline src/ui/components/search/autocompletecomponent.js]: levenshtein

File: src/ui/components/search/autocompletecomponent.js [hunk 4]
```diff
@@ -179,6 +187,19 @@ export default class AutoCompleteComponent extends Component {
...
+        isCloseMatch(prompt, queryInputEl.value)
+      );
+      if (!hasCustomPrompts && matchedCustomPrompts.length > 0) {
+        const dataWithCustomPrompts = this.combineResultsWithCustomPrompts(data, matchedCustomPrompts);
+        this.core.storage.set(`${StorageKeys.AUTOCOMPLETE}.${this.name}`, dataWithCustomPrompts);
+        return;
+      }
+    }
+
     if (wasOpen && !this._isOpen) {
```
Comment (oshi97) [inline src/ui/components/search/autocompletecomponent.js:203]: Do we need to worry about this wasOpen and isOpen stuff for custom prompts?

File: src/ui/tools/stringmatching.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import levenshtein from 'js-levenshtein';
```
Comment (oshi97) [inline src/ui/tools/stringmatching.js:1]: Should we update the FilterOptions component to use these functions? For a bit of history, the whole levenshtein stuff we do actually originated there from when Alexis did searchable facets. It's kind of funny how it's come full circle. I don't think it should be too hard to update and test, but if it is I think it's totally fine if we say it's not worth the developer time.

File: src/ui/components/search/autocompletecomponent.js [hunk 6]
```diff
@@ -358,6 +379,28 @@ export default class AutoCompleteComponent extends Component {
     this._isVoiceSearchActive = value;
   }
 
+  setCustomPrompts (prompts) {
+    this.customPrompts = prompts;
+  }
+
+  createSectionForCustomPrompts (customPrompts) {
+    const customAutocompleteResults = customPrompts
+      .map(prompt => AutoCompleteResponseTransformer._transformAutoCompleteResult({ value: prompt }));
```
Comment (oshi97) [inline src/ui/components/search/autocompletecomponent.js]: Could we remove the underscore from `_transformAutoCompleteResult` if it's being used "publicly" now?

File: src/ui/components/componentmanager.js [hunk 7]
```diff
@@ -200,6 +200,10 @@ export default class ComponentManager {
     return this._activeComponents.find(c => c.constructor.type === type);
   }
 
+  getActiveComponents (type) {
```
Comment (oshi97) [inline src/ui/components/componentmanager.js:203]: For my edification how is this being used? I didn't see any usages in the fresh code

File: src/ui/components/search/autocompletecomponent.js [hunk 8]
```diff
@@ -179,6 +187,19 @@ export default class AutoCompleteComponent extends Component {
...
 
+    // show custom prompts if it's an empty search or levenshtein match on a non empty search
+    if (this.customPrompts && data.sections) {
+      const hasCustomPrompts = data.sections.find(result => result.isCustom);
+      const matchedCustomPrompts = this.customPrompts.filter(prompt =>
+        isCloseMatch(prompt, queryInputEl.value)
+      );
+      if (!hasCustomPrompts && matchedCustomPrompts.length > 0) {
+        const dataWithCustomPrompts = this.combineResultsWithCustomPrompts(data, matchedCustomPrompts);
+        this.core.storage.set(`${StorageKeys.AUTOCOMPLETE}.${this.name}`, dataWithCustomPrompts);
```
Comment (oshi97) [inline src/ui/components/search/autocompletecomponent.js:198]: Is this kind of like doing a useLayoutEffect in react, where we update the state without a paint and then trigger a re-render?

File: src/ui/components/componentmanager.js (hunk 7 above)
Comment (oshi97) [reply src/ui/components/componentmanager.js:203]: Ah ok I see it in the hitchhiker-theme PR now

File: src/ui/tools/stringmatching.js (hunk 5 above)
Comment (yen-tt) [reply src/ui/tools/stringmatching.js:1]: There are some small differences with FilterOptions (such as needing the offset of substrings in the search term) that would require a bit of changes. I think I would prefer if we make another item/do it in another pr for this, if that's alright?

File: src/ui/components/search/autocompletecomponent.js (hunk 6 above)
Comment (yen-tt) [reply src/ui/components/search/autocompletecomponent.js]: sure!

File: src/ui/components/search/autocompletecomponent.js (hunk 8 above)
Comment (yen-tt) [reply src/ui/components/search/autocompletecomponent.js:198]: yea, we update the state to include the custom prompts section (this is needed so keyboard navigation and query submission registers the right sections and indexes), and the return statement following this line ensures the old data doesn't get render.

File: src/ui/tools/stringmatching.js (hunk 5 above)
Comment (oshi97) [reply src/ui/tools/stringmatching.js:1]: 👍  yeah!

**Review by oshi97 — APPROVED** [review]

**Pull request merged**